src/vos/vos_ts.h

```diff
@@ -760,6 +813,46 @@ vos_ts_set_update(struct vos_ts_set *ts_set, daos_epoch_t read_time)
 	}
 }
 
+static inline void
+vos_ts_update_wcache(struct vos_wts_cache *wcache, daos_epoch_t write_time)
+{
+	daos_epoch_t		*max;
+	daos_epoch_t		*min;
+
+	max = &wcache->wc_ts_w[wcache->wc_w_max];
+	min = &wcache->wc_ts_w[1 - wcache->wc_w_max];
+
+	if (write_time <= *min || write_time == *max)
+		return;
+
+	if (write_time > *max)
+		wcache->wc_w_max = 1 - wcache->wc_w_max;
+	*min = write_time;
+}
+
+/** Update the write timestamps for the set after a successful operation
+ *
+ *  \param[in]	ts_set		The timestamp set
+ *  \param[in]	write_time	The new write timestamp
+ */
+static inline void
+vos_ts_set_wupdate(struct vos_ts_set *ts_set, daos_epoch_t write_time)
+{
+	struct vos_ts_set_entry	*se;
+	int			 i;
+
+	if (!vos_ts_in_tx(ts_set))
+		return;
```

Comment:
An epoch uncertainty check also looks for standalone writes. Yeah, we should add those cases to vts_mvcc.

Reply:
There are no such cases in practice.   Even solo writes pass in a DTX so from VOS perspective, they look like a transaction.

Reply:
Yes, indeed. Sorry.